pipeline_branching.md

```diff
@@ -0,0 +1,159 @@
+# Pipeline Branching
+
+## Goal
+
+Allow users to have multiple, named versions of a pipeline (branches)
+running concurrently.
+
+## Use Case:
+
+The primary use case for this is to give users a good way to have `dev`,
+`staging` and `prod` versions of pipelines. There are a ton of use cases for
+this feature, but so far I haven't encountered one that presented challenges
+that the `dev`, `staging` `prod` use case doesn't, so for this document we'll
+just stick to that for inspiration.
+
+## PPS API changes:
+The pipeline message will be extended to include a `Branch` field, which
+is a string:
+
+```proto
+message Pipeline {
+  string name = 1;
+  string branch = 2;
+}
+```
+
+In pachctl this is implemented with the familiar syntax of `pipeline@branch`.
+
+An empty `branch` field is equivalent of a value of `"master"`.
+
+Different branches of the same pipeline are effectively different pipelines. That means:
```

Comment:
For my own clarity: Over time master and branch can diverge significantly -- that's a valid DAG state?

Reply:
Yeah they can diverge basically as much as the user wants. I expect that most of the time the change will just be a different Transform with a new version of the code. Less frequent but still totally valid is changing what inputs it takes, i.e. changing the shape of the DAG. At a certain point, if you're going to change everything then it probably makes more sense to just make a totally new DAG rather than branching. Since branching is a tool for iterating, if there's nothing that's staying the same it's just going to add extra complexity.